File: shoebill_plugins/EntrancePlugin/src/main/java/lt/ltrp/object/Entrance.java
package lt.ltrp.object;

import lt.ltrp.EntrancePlugin;
import lt.maze.streamer.object.DynamicLabel;
import lt.maze.streamer.object.DynamicPickup;
import net.gtaun.shoebill.data.AngledLocation;
import net.gtaun.shoebill.data.Color;
import net.gtaun.shoebill.data.Location;

import java.util.Collection;

/**
 * @author Bebras
 *         2016.05.22.
 */
public interface Entrance extends DestroyableEntity {

    static final int DEFAULT_PICKUP_MODEL = 19606;
    static final Color DEFAULT_LABEL_COLOR = new Color(0x1299A2AA);
    static final int MIN_TEXT_LENGTH = 10;

    static Entrance create(AngledLocation location, String name) {
        return create(location, name, DEFAULT_LABEL_COLOR, DEFAULT_PICKUP_MODEL, null, false);
    }

    static Entrance create(AngledLocation location, String text, Color labelColor, int pickupModelId, Job job, boolean allowsVehicles) {
        return EntrancePlugin.get(EntrancePlugin.class).createEntrance(location, text, labelColor, pickupModelId, job, allowsVehicles);
    }

    static Collection<Entrance> get() {
        return EntrancePlugin.get(EntrancePlugin.class).getEntrances();
    }

    static Entrance get(int uuid) {
        return EntrancePlugin.get(EntrancePlugin.class).get(uuid);
    }

    static Entrance getClosest(Location location) {
        return getClosest(location, Float.POSITIVE_INFINITY);
    }

    static Entrance getClosest(Location location, float maxDistance) {
        return EntrancePlugin.get(EntrancePlugin.class).getClosest(location, maxDistance);
    }


    AngledLocation getLocation();
    void setLocation(AngledLocation angledLocation);

    DynamicPickup getPickup();
    DynamicLabel getLabel();

    String getText();
    void setText(String text);

    int getPickupModelId();
    void setPickupModelId(int modelId);

    Color getColor();
    void setColor(Color color);

    boolean allowsVehicles();
    void setAllowVehicles(boolean set);

    boolean isJob();
    Job getJob();
    void setJob(Job job);

    Entrance getExit();
    void setExit(Entrance entrance);

}

File: shoebill_plugins/EntrancePlugin/src/main/java/lt/ltrp/entrance/object/Entrance.kt
package lt.ltrp.entrance.`object`

import lt.ltrp.entrance.EntrancePlugin
import lt.ltrp.job.`object`.Job
import lt.ltrp.`object`.DestroyableEntity
import lt.maze.streamer.`object`.DynamicLabel
import lt.maze.streamer.`object`.DynamicPickup
import net.gtaun.shoebill.data.AngledLocation
import net.gtaun.shoebill.data.Color
import net.gtaun.shoebill.data.Location

/**
 * @author Bebras
 * *         2016.05.22.
 */
interface Entrance : DestroyableEntity {


    var location: AngledLocation

    val pickup: DynamicPickup
    val label: DynamicLabel

    var text: String

    var pickupModelId: Int

    var color: Color

    fun allowsVehicles(): Boolean
    fun setAllowVehicles(set: Boolean)

    val isJob: Boolean
    var job: Job

    var exit: Entrance

    companion object {

        val DEFAULT_PICKUP_MODEL = 19606
        val DEFAULT_LABEL_COLOR = Color(0x1299A2AA)
        val MIN_TEXT_LENGTH = 10

        fun create(location: AngledLocation, name: String): Entrance {
            return create(location, name, DEFAULT_LABEL_COLOR, DEFAULT_PICKUP_MODEL, null, false)
        }

        fun create(location: AngledLocation, text: String, labelColor: Color, pickupModelId: Int, job: Job?, allowsVehicles: Boolean): Entrance {
            return EntrancePlugin.get(EntrancePlugin::class.java.toInt()).createEntrance(location, text, labelColor, pickupModelId, job, allowsVehicles)
        }

        fun get(): Collection<Entrance> {
            return EntrancePlugin.get(EntrancePlugin::class.java.toInt()).getEntrances()
        }

        operator fun get(uuid: Int): Entrance {
            return EntrancePlugin.get(EntrancePlugin::class.java.toInt()).get(uuid)
        }

        @JvmOverloads fun getClosest(location: Location, maxDistance: Float = java.lang.Float.POSITIVE_INFINITY): Entrance {
            return EntrancePlugin.get(EntrancePlugin::class.java.toInt()).getClosest(location, maxDistance)
        }
    }

}
